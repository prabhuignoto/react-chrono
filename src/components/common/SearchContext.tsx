import React, { createContext, useState, useContext } from 'react';

interface SearchContextType {
  searchTerm: string;
  setSearchTerm: (term: string) => void;
}

// Create a context with default values
const SearchContext = createContext<SearchContextType>({
  searchTerm: '',
  setSearchTerm: () => {},
});

/**
 * Provider component for the search context
 */
export const SearchProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [searchTerm, setSearchTerm] = useState<string>('');

  return (
    <SearchContext.Provider value={{ searchTerm, setSearchTerm }}>
      {children}
    </SearchContext.Provider>
  );
};

/**
 * Custom hook to use the search context
 */
export const useSearch = (): SearchContextType => {
  const context = useContext(SearchContext);

  if (context === undefined) {
    throw new Error('useSearch must be used within a SearchProvider');
  }

  return context;
};

export default SearchContext;
