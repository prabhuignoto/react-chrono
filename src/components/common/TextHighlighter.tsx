import React from 'react';
import styled from 'styled-components';
import { Theme } from '@models/Theme';

interface HighlightedTextProps {
  text: string;
  searchTerm: string;
  theme: Theme;
}

// Styled component for highlighted text
const HighlightSpan = styled.span<{ theme: Theme }>`
  background-color: ${({ theme }) =>
    theme.primary}33; /* Primary color with 20% opacity */
  color: ${({ theme }) => theme.primary};
  font-weight: 500;
  border-radius: 2px;
  padding: 0 2px;
`;

/**
 * Component that highlights occurrences of a search term within text
 */
const TextHighlighter: React.FC<HighlightedTextProps> = ({
  text,
  searchTerm,
  theme,
}) => {
  // If no search term or no text, just return the original text
  if (!searchTerm || !text) {
    return <>{text}</>;
  }

  // Escape special regex characters in the search term
  const escapedSearchTerm = searchTerm.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  // Create a case-insensitive regular expression to find all occurrences
  const regex = new RegExp(`(${escapedSearchTerm})`, 'gi');

  // Split the text by the search term
  const parts = text.split(regex);

  // Return the text with highlighted parts
  return (
    <>
      {parts.map((part, i) =>
        regex.test(part) ? (
          <HighlightSpan key={i} theme={theme}>
            {part}
          </HighlightSpan>
        ) : (
          <React.Fragment key={i}>{part}</React.Fragment>
        ),
      )}
    </>
  );
};

export default TextHighlighter;
