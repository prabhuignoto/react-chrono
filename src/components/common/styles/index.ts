import { Theme } from '@models/Theme';
import { css } from 'styled-components';

export const ScrollBar = css<{ theme: Theme }>`
  scrollbar-color: ${(p) => p.theme?.primary} default;
  scrollbar-width: thin;

  &::-webkit-scrollbar {
    width: 0.3em;
  }

  &::-webkit-scrollbar-track {
    box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.2);
  }

  &::-webkit-scrollbar-thumb {
    background-color: ${(p) => p.theme?.primary};
    outline: 1px solid ${(p) => p.theme?.primary};
  }
`;
