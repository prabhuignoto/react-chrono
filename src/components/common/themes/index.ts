import { Theme } from '@models/Theme';

export const defaultTheme: Theme = {
  cardBgColor: '#ffffff',
  cardDetailsBackGround: '#ffffff',
  cardDetailsColor: '#000',
  cardMediaBgColor: '#f5f5f5',
  cardSubtitleColor: '#000',
  cardTitleColor: '#007FFF',
  detailsColor: '#000',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#f5f5f5',
  nestedCardDetailsBackGround: '#f5f5f5',
  nestedCardDetailsColor: '#000',
  nestedCardSubtitleColor: '#000',
  nestedCardTitleColor: '#000',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#f5f5f5',
  toolbarBtnBgColor: '#fff',
  toolbarTextColor: '#000',
};

export const darkTheme: Theme = {
  cardBgColor: '#191919',
  cardDetailsBackGround: '#191919',
  cardDetailsColor: '#ffff0f',
  cardMediaBgColor: '#2f2f2f',
  cardSubtitleColor: '#ffffff',
  cardTitleColor: '#007FFF',
  detailsColor: '#ffffff',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#333333',
  nestedCardDetailsBackGround: '#333333',
  nestedCardDetailsColor: '#ffffff',
  nestedCardSubtitleColor: '#ffffff',
  nestedCardTitleColor: '#ffffff',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#555',
  toolbarBtnBgColor: '#222',
  toolbarTextColor: '#ffffff',
};
