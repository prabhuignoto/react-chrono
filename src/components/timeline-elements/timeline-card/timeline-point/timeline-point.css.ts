import { keyframes, style, globalStyle } from '@vanilla-extract/css';
import { recipe } from '@vanilla-extract/recipes';
import { vars } from '../../../../styles/tokens.css';
import { sprinkles } from '../../../../styles/sprinkles/sprinkles.css';

// Keyframe animations
const ripple = keyframes({
  '0%': { transform: 'scale(0)', opacity: 0.8 },
  '100%': { transform: 'scale(4)', opacity: 0 },
});

const pulse = keyframes({
  '0%': { boxShadow: '0 0 0 0 rgba(0, 123, 255, 0.6)' },
  '70%': { boxShadow: '0 0 0 12px rgba(0, 123, 255, 0)' },
  '100%': { boxShadow: '0 0 0 0 rgba(0, 123, 255, 0)' },
});

export const shapeWrapper = style([
  sprinkles({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  }),
  {
    width: 'auto', // Let content determine width
    height: 'auto', // Let content determine height
    position: 'relative',
    // Simplified - no complex z-index or excessive padding
    flexShrink: 0,
  },
]);

export const timelinePointBase = style({
  cursor: 'pointer',
  transition: 'all 0.2s ease-in-out',
  position: 'relative',
  overflow: 'visible',
  padding: 0,
  width: '1.5rem', // Smaller size for horizontal modes
  height: '1.5rem',
  minWidth: '1.5rem',
  minHeight: '1.5rem',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',

  // Ripple effect
  '::before': {
    content: '""',
    position: 'absolute',
    top: '50%',
    left: '50%',
    width: '10px',
    height: '10px',
    backgroundColor: 'rgba(255, 255, 255, 0.5)',
    borderRadius: '50%',
    transform: 'scale(0) translate(-50%, -50%)',
    transformOrigin: 'top left',
    pointerEvents: 'none',
    zIndex: 3,
    opacity: 0,
  },

  selectors: {
    '&:active::before': {
      animation: `${ripple} 0.6s ease-out`,
    },
    '&:focus-visible': {
      outline: '3px solid transparent',
      outlineOffset: '2px',
    },
    '&:disabled': {
      opacity: 0.6,
      cursor: 'default',
    },
  },

  '@media': {
    '(prefers-reduced-motion: reduce)': {
      animation: 'none',
      transition: 'none',
    },
  },
});

export const timelinePoint = recipe({
  base: timelinePointBase,
  variants: {
    shape: {
      circle: {
        borderRadius: '50%',
        // Ensure circle shape is always visible
        backgroundColor: 'currentColor',
        border: '2px solid white',
      },
      square: {
        borderRadius: '2px',
        backgroundColor: 'currentColor',
        border: '2px solid white',
      },
      diamond: {
        borderRadius: '0',
        transform: 'rotate(45deg)',
        backgroundColor: 'currentColor',
        border: '2px solid white',
      },
    },
    usingIcon: {
      true: {
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
      },
      false: {},
    },
    active: {
      true: {
        transform: 'scale(1.2)', // Slightly smaller scale for smaller points
        transition: 'all 0.3s ease-in-out',
        backgroundColor: 'currentColor',
        border: '2px solid white',
        boxShadow: '0 3px 10px rgba(0, 123, 255, 0.4)',
        selectors: {
          '&.using-icon': {
            animation: `${pulse} 1.5s infinite`,
          },
          '&:not(.using-icon)': {
            boxShadow: '0 0 10px rgba(0, 123, 255, 0.6)',
            animation: `${pulse} 1.5s infinite`,
          },
        },
      },
      false: {
        transform: 'scale(1)',
        transition: 'all 0.3s ease-in-out',
        // Ensure timeline points are always visible with strong styling
        backgroundColor: 'currentColor',
        border: '2px solid white',
        boxShadow: '0 2px 6px rgba(0, 0, 0, 0.15)',
        // Add hover effect for non-active points
        selectors: {
          '&:hover:not(:disabled)': {
            transform: 'scale(1.15)',
            boxShadow: '0 3px 10px rgba(0, 0, 0, 0.2)',
          },
        },
      },
    },
    disabled: {
      true: {
        opacity: 0.6,
        cursor: 'default',
      },
      false: {},
    },
  },
  compoundVariants: [
    {
      variants: { shape: 'diamond', active: false },
      style: {
        selectors: {
          '&:hover:not(:disabled)': {
            transform: 'rotate(45deg) scale(1.12)',
          },
          '&:active:not(:disabled)': {
            transform: 'rotate(45deg) scale(0.98)',
          },
        },
      },
    },
    {
      variants: { shape: 'diamond', active: true, usingIcon: true },
      style: {
        transform: 'rotate(45deg) scale(1.1)',
      },
    },
    {
      variants: { shape: 'diamond', active: true, usingIcon: false },
      style: {
        transform: 'rotate(45deg)',
      },
    },
  ],
  defaultVariants: {
    shape: 'circle',
    usingIcon: false,
    active: false,
    disabled: false,
  },
});

// Global styles for icon images
globalStyle(`${timelinePointBase}.using-icon img`, {
  maxWidth: '90%',
  maxHeight: '90%',
});
