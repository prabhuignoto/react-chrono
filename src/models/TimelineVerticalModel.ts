import { Theme } from './Theme';
import { Scroll } from './TimelineHorizontalModel';
import { TimelineCardModel } from './TimelineItemModel';
import { Media } from './TimelineMediaModel';
import { TimelineProps } from './TimelineModel';

export type Props = Pick<
  TimelineProps,
  | 'flipLayout'
  | 'theme'
  | 'mode'
  | 'timelineCircleDimension'
  | 'lineWidth'
  | 'cardHeight'
  | 'enableOutline'
  | 'disableClickOnCircle'
  | 'cardLess'
> & {
  alternateCards?: boolean;
  hasFocus?: boolean;
  onClick: (id?: string) => void;
  onElapsed?: (id?: string) => void;
  slideItemDuration?: number;
  slideShowRunning?: boolean;
  theme?: Theme;
};

type VerticalModel = Pick<
  Props,
  | 'alternateCards'
  | 'hasFocus'
  | 'onClick'
  | 'onElapsed'
  | 'slideShowRunning'
  | 'mode'
  | 'timelineCircleDimension'
  | 'lineWidth'
  | 'disableClickOnCircle'
  | 'cardLess'
> & {
  active?: boolean;
  className: string;
  id?: string;
};

export interface VerticalCircleModel extends VerticalModel {
  iconChild?: React.ReactNode;
  onActive: (pointOffset: number) => void;
}

export interface VerticalItemModel extends VerticalModel {
  cardDetailedText?: string | string[];
  cardSubtitle?: string;
  cardTitle?: string;
  contentDetailsChildren?: React.ReactNode;
  iconChild?: React.ReactNode;
  index: number;
  media?: Media;
  onActive: (
    pointOffset: number,
    contentHeight: number,
    contentOffset: number,
  ) => void;
  onShowMore?: () => void;
  title?: string;
  url?: string;
  visible?: boolean;
}

export type TimelineVerticalModel = Pick<
  Props,
  | 'alternateCards'
  | 'enableOutline'
  | 'mode'
  | 'onClick'
  | 'onElapsed'
  | 'slideShowRunning'
  | 'theme'
  | 'hasFocus'
  | 'cardLess'
> & {
  activeTimelineItem: number;
  autoScroll: (s: Partial<Scroll>) => void;
  childrenNode?: React.ReactNode;
  contentDetailsChildren?: React.ReactNode;
  iconChildren?: React.ReactNode;
  items: TimelineCardModel[];
  onOutlineSelection?: (index: number) => void;
};
