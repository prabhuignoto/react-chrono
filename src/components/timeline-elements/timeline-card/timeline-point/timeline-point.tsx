import React from 'react';
import { Shape, ShapeWrapper } from '../timeline-horizontal-card.styles';

interface TimelinePointProps {
  circleClass: string;
  handleClick: () => void;
  circleRef: React.RefObject<HTMLButtonElement>;
  title?: string;
  theme?: any;
  timelinePointDimension?: number;
  timelinePointShape?: 'circle' | 'square' | 'diamond';
  iconChild?: React.ReactNode;
  active?: boolean;
  disabled?: boolean;
}

const TimelinePoint: React.FC<TimelinePointProps> = ({
  circleClass,
  handleClick,
  circleRef,
  title,
  theme,
  timelinePointDimension,
  timelinePointShape,
  iconChild,
  active = false,
  disabled = false,
}) => {
  return (
    <ShapeWrapper>
      <Shape
        as="button"
        className={circleClass}
        onClick={handleClick}
        ref={circleRef}
        data-testid="timeline-circle"
        theme={theme}
        aria-label={title ?? 'Timeline point'}
        aria-selected={active}
        aria-disabled={disabled}
        disabled={disabled}
        dimension={timelinePointDimension}
        $timelinePointShape={timelinePointShape}
        tabIndex={disabled ? -1 : 0}
        type="button"
      >
        {iconChild ?? null}
      </Shape>
    </ShapeWrapper>
  ) as React.ReactElement;
};

export default TimelinePoint;
