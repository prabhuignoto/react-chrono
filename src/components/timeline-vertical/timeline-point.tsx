import { TimelinePointModel } from '@models/TimelineVerticalModel';
import cls from 'classnames';
import React, { memo, useContext, useEffect, useMemo, useRef } from 'react';
import { GlobalContext } from '../GlobalContext';
import { Shape } from '../timeline-elements/timeline-card/timeline-horizontal-card.styles';
import {
  TimelinePointContainer,
  TimelinePointWrapper,
} from './timeline-vertical-shape.styles';

/**
 * TimelinePoint
 * @property {string} className - The class name for the component.
 * @property {string} id - The id of the timeline point.
 * @property {() => void} onClick - Function to handle click event.
 * @property {boolean} active - Whether the timeline point is active.
 * @property {() => void} onActive - Function to handle active event.
 * @property {boolean} slideShowRunning - Whether the slideshow is running.
 * @property {React.ReactNode} iconChild - The icon child nodes.
 * @property {number} timelinePointDimension - The dimension of the timeline point.
 * @property {number} lineWidth - The width of the line.
 * @property {boolean} disableClickOnCircle - Whether the click on circle is disabled.
 * @property {boolean} cardLess - Whether the card is less.
 * @returns {JSX.Element} The TimelinePoint component.
 */
const TimelinePoint: React.FunctionComponent<TimelinePointModel> = memo(
  (props: TimelinePointModel) => {
    const {
      className,
      id,
      onClick,
      active,
      onActive,
      slideShowRunning,
      iconChild,
      timelinePointDimension,
      lineWidth,
      disableClickOnCircle,
      cardLess,
      isMobile,
    } = props;

    const circleRef = useRef<HTMLButtonElement>(null);
    const {
      theme,
      focusActiveItemOnLoad,
      timelinePointShape,
      disableTimelinePoint,
    } = useContext(GlobalContext);

    const isFirstRender = useRef(true);

    // Determine if onActive can be invoked
    const canInvokeOnActive = useMemo(() => {
      if (focusActiveItemOnLoad) {
        return active;
      } else {
        return active && !isFirstRender.current;
      }
    }, [active]);

    // Invoke onActive if conditions are met
    useEffect(() => {
      if (canInvokeOnActive) {
        const circle = circleRef.current;

        circle && onActive(circle.offsetTop);
      }
    }, [canInvokeOnActive, active]);

    // Determine circle class
    const circleClass = useMemo(
      () =>
        cls({
          active,
          'using-icon': !!iconChild,
        }),
      [active, iconChild],
    );

    // Determine click handler props
    const clickHandlerProps = useMemo(
      () =>
        !disableClickOnCircle && {
          onClick: (ev: React.MouseEvent) => {
            ev.stopPropagation();
            if (id && onClick && !slideShowRunning) {
              onClick(id);
            }
          },
        },
      [id, onClick, slideShowRunning, disableClickOnCircle],
    );

    // Update isFirstRender flag after first render
    useEffect(() => {
      if (isFirstRender.current) {
        isFirstRender.current = false;
      }
    }, []);

    return (
      <TimelinePointWrapper
        width={lineWidth}
        bg={theme && theme.primary}
        className={className}
        data-testid="tree-leaf"
        $cardLess={cardLess}
        $isMobile={isMobile}
      >
        {/* {!disableTimelinePoint ? ( */}
        <TimelinePointContainer
          className={`${className} timeline-vertical-circle`}
          {...clickHandlerProps}
          ref={circleRef}
          data-testid="tree-leaf-click"
          aria-label="select timeline"
          $hide={disableTimelinePoint}
        >
          <Shape
            className={circleClass}
            theme={theme}
            dimension={timelinePointDimension}
            $timelinePointShape={timelinePointShape}
          >
            {iconChild ? iconChild : null}
          </Shape>
        </TimelinePointContainer>
        {/* ) : null} */}
      </TimelinePointWrapper>
    );
  },
  (prev, next) =>
    prev.active === next.active && prev.isMobile === next.isMobile,
);

TimelinePoint.displayName = 'TimelinePoint';

export { TimelinePoint };
