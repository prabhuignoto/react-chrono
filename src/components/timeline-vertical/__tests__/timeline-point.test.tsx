import { TimelinePointModel } from '@models/TimelineVerticalModel';
import { describe, expect, it, vi } from 'vitest';
import { customRender } from '../../common/test';
import { providerProps } from '../../common/test/index';
import { TimelinePoint } from '../timeline-point';

const commonProps: TimelinePointModel = {
  active: false,
  alternateCards: false,
  cardLess: true,
  className: 'test_class_name',
  disableClickOnCircle: false,
  iconChild: null,
  id: '1',
  lineWidth: 3,
  onActive: vi.fn(),
  onClick: vi.fn(),
  slideShowRunning: false,
  timelinePointDimension: 20,
};

describe('Timeline point', () => {
  // should match the snapshot

  it('Should match snapshot', () => {
    const { asFragment } = customRender(<TimelinePoint {...commonProps} />, {
      providerProps,
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
