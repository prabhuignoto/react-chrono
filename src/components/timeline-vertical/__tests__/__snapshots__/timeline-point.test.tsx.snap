// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Timeline point > Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gLDzan KfHIG test_class_name"
    data-testid="tree-leaf"
    role="button"
    width="3"
  >
    <div
      aria-label="select timeline"
      class="sc-iAEyYk iOfSrA test_class_name timeline-vertical-circle"
      data-testid="tree-leaf-click"
      role="button"
    >
      <div
        class="sc-hLseeU iQJQbg"
      />
    </div>
  </div>
</DocumentFragment>
`;
